File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/JwtSigned.Extensions.vb
Imports System
Imports JWT
Imports JWT.Algorithms
Imports JWT.Builder
Imports System.Collections.Generic

Namespace Tests.Diagnostics

    Class Program

        Protected Const Secret As String = "GQDstcKsx0NHjPOuXOYg5MbeJ1XT0uFiwDVvVBrk"
        Protected Const InvalidToken As String = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJmb28iOiJmYWtlYmFyIiwiaWF0IjoxNTc1NjQ0NTc3fQ.pcX_7snpSGf01uBfaM8XPkbgdhs1gq9JcYRCQvZrJyk"

        Private InvalidParts As JwtParts

        Sub ExtensionCalledOnInstance(Decoder As JwtDecoder)

            Dim byteArray As Byte() = { 16, 23 }

            Dim Decoded As String = Decoder.Decode(invalidToken) ' Noncompliant
            Decoded = Decoder.Decode(invalidParts) ' Noncompliant

            Decoded = Decoder.Decode(invalidToken, byteArray, false) ' Noncompliant
            Decoded = Decoder.Decode(invalidToken, byteArray, true) ' Compliant

            Decoded = Decoder.Decode(invalidToken, secret, false) ' Noncompliant
            Decoded = Decoder.Decode(invalidToken, secret, true) ' Compliant

            Dim DecodedObject As Object = Decoder.DecodeToObject(Of Object)(invalidToken, byteArray, false) ' Noncompliant
            DecodedObject = Decoder.DecodeToObject(Of Object)(invalidToken, byteArray, true) ' Compliant

            DecodedObject = Decoder.DecodeToObject(Of Object)(invalidToken, secret, false) ' Noncompliant
            DecodedObject = Decoder.DecodeToObject(Of Object)(invalidToken, secret, true) ' Compliant

            DecodedObject = Decoder.DecodeToObject(Of Object)(invalidToken) ' Noncompliant
            Dim DecodedMap As IDictionary(Of String, Object) = decoder.DecodeToObject(invalidToken) ' Noncompliant

        End Sub

        Sub ExtensionCalledThroughStaticClass(Decoder As JwtDecoder)

            Dim byteArray As Byte() = { 16, 23 }

            Dim Decoded As String = JwtDecoderExtensions.Decode(Decoder, invalidToken) ' Noncompliant
            Decoded = JwtDecoderExtensions.Decode(Decoder, invalidParts) ' Noncompliant

            Decoded = JwtDecoderExtensions.Decode(Decoder, invalidToken, byteArray, false) ' Noncompliant
            Decoded = JwtDecoderExtensions.Decode(Decoder, invalidToken, byteArray, true) ' Compliant

            Decoded = JwtDecoderExtensions.Decode(Decoder, invalidToken, secret, false) ' Noncompliant
            Decoded = JwtDecoderExtensions.Decode(Decoder, invalidToken, secret, true) ' Compliant

            Dim DecodedObject As Object = JwtDecoderExtensions.DecodeToObject(Of Object)(Decoder, invalidToken, byteArray, false) ' Noncompliant
            DecodedObject = JwtDecoderExtensions.DecodeToObject(Of Object)(Decoder, invalidToken, byteArray, true) ' Compliant

            DecodedObject = JwtDecoderExtensions.DecodeToObject(Of Object)(Decoder, invalidToken, secret, false) ' Noncompliant
            DecodedObject = JwtDecoderExtensions.DecodeToObject(Of Object)(Decoder, invalidToken, secret, true) ' Compliant

            DecodedObject = JwtDecoderExtensions.DecodeToObject(Of Object)(Decoder, invalidToken) ' Noncompliant
            Dim DecodedMap As IDictionary(Of String, Object) = JwtDecoderExtensions.DecodeToObject(Decoder, invalidToken) ' Noncompliant

        End Sub

    End Class

End Namespace

File: analyzers/tests/SonarAnalyzer.Rules.Tests/Cases/JwtSigned.Extensions.cs
using System;
using System.Collections.Generic;
using JWT;
using JWT.Algorithms;
using JWT.Builder;

namespace Tests.Diagnostics
{
    class Program
    {
        const string secret = "GQDstcKsx0NHjPOuXOYg5MbeJ1XT0uFiwDVvVBrk";
        const string invalidToken = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJmb28iOiJmYWtlYmFyIiwiaWF0IjoxNTc1NjQ0NTc3fQ.pcX_7snpSGf01uBfaM8XPkbgdhs1gq9JcYRCQvZrJyk";

        private JwtParts invalidParts;

        void ExtensionCalledOnInstance(JwtDecoder decoder)
        {
            var decoded = decoder.Decode(invalidToken); // Noncompliant
            decoded = decoder.Decode(invalidParts); // Noncompliant

            decoded = decoder.Decode(invalidToken, new byte[] { 42 }, false); // Noncompliant
            decoded = decoder.Decode(invalidToken, new byte[] { 42 }, true); // Compliant

            decoded = decoder.Decode(invalidToken, secret, false); // Noncompliant
            decoded = decoder.Decode(invalidToken, secret, true); // Compliant

            var decodedObject = decoder.DecodeToObject<object>(invalidToken, new byte[] { 42 }, false); // Noncompliant
            decodedObject = decoder.DecodeToObject<object>(invalidToken, new byte[] { 42 }, true); // Compliant

            decodedObject = decoder.DecodeToObject<object>(invalidToken, secret, false); // Noncompliant
            decodedObject = decoder.DecodeToObject<object>(invalidToken, secret, true); // Compliant

            decodedObject = decoder.DecodeToObject<object>(invalidToken); // Noncompliant
            IDictionary<string, object> decodedMap = decoder.DecodeToObject(invalidToken); // Noncompliant
        }

        void ExtensionCalledThroughStaticClass(JwtDecoder decoder)
        {
            var decoded = JwtDecoderExtensions.Decode(decoder, invalidToken); // Noncompliant
            decoded = JwtDecoderExtensions.Decode(decoder, invalidParts); // Noncompliant

            decoded = JwtDecoderExtensions.Decode(decoder, invalidToken, new byte[] { 42 }, false); // Noncompliant
            decoded = JwtDecoderExtensions.Decode(decoder, invalidToken, new byte[] { 42 }, true); // Compliant

            decoded = JwtDecoderExtensions.Decode(decoder, invalidToken, secret, false); // Noncompliant
            decoded = JwtDecoderExtensions.Decode(decoder, invalidToken, secret, true); // Compliant

            var decodedObject = JwtDecoderExtensions.DecodeToObject<object>(decoder, invalidToken, new byte[] { 42 }, false); // Noncompliant
            decodedObject = JwtDecoderExtensions.DecodeToObject<object>(decoder, invalidToken, new byte[] { 42 }, true); // Compliant

            decodedObject = JwtDecoderExtensions.DecodeToObject<object>(decoder, invalidToken, secret, false); // Noncompliant
            decodedObject = JwtDecoderExtensions.DecodeToObject<object>(decoder, invalidToken, secret, true); // Compliant

            decodedObject = JwtDecoderExtensions.DecodeToObject<object>(decoder, invalidToken); // Noncompliant
            IDictionary<string, object> decodedMap = JwtDecoderExtensions.DecodeToObject(decoder, invalidToken); // Noncompliant
        }
    }
}
